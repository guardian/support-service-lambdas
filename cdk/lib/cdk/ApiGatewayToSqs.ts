import { AwsIntegration, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import type { Queue } from 'aws-cdk-lib/aws-sqs';
import { Construct } from 'constructs';
import { SrApiGateway5xxAlarm } from './SrApiGateway5xxAlarm';
import type { SrMonitoring } from './SrLambdaAlarm';
import type { SrStack } from './SrStack';

type ApiGatewayToSqsProps = {
	queue: Queue;
	/**
	 * header names to pull from the request into SQS attributes.
	 *
	 * FIXME if a header is missing, the requester still gets a 200 and the message is silently dropped!
	 */
	includeHeaderNames: string[];
	/**
	 * do we want to disable standard SrCDK 5xx alarm or override any properties?
	 */
	monitoring: SrMonitoring;
};

/**
 * this takes incoming http requests and puts them on a queue, responding with a 200 OK
 *
 * Useful for webhooks etc.
 */
export class ApiGatewayToSqs extends Construct {
	constructor(scope: SrStack, id: string, props: ApiGatewayToSqsProps) {
		super(scope, id);

		const apiRole = new Role(this, 'Role', {
			assumedBy: new ServicePrincipal('apigateway.amazonaws.com'),
		});
		props.queue.grantSendMessages(apiRole);

		const sendMessageIntegration = new AwsIntegration({
			service: 'sqs',
			path: `${scope.account}/${props.queue.queueName}`,
			integrationHttpMethod: 'POST',
			options: {
				credentialsRole: apiRole,
				requestParameters: {
					'integration.request.header.Content-Type':
						"'application/x-www-form-urlencoded'",
				},
				requestTemplates: {
					'application/json': [
						...Object.entries({
							Action: 'SendMessage',
							MessageBody: '$util.urlEncode($input.body)',
						}),
						...props.includeHeaderNames.flatMap((headerName, index) => [
							[`MessageAttribute.${index + 1}.Name`, headerName],
							[`MessageAttribute.${index + 1}.Value.DataType`, 'String'],
							[
								`MessageAttribute.${index + 1}.Value.StringValue`,
								`$method.request.header.${headerName}`,
							],
						]),
					]
						.map(([k, v]) => k + '=' + v)
						.join('&'),
				},
				integrationResponses: [
					{
						statusCode: '200',
						responseTemplates: {
							'application/json': '{ "status": "accepted" }',
						},
					},
				],
			},
		});

		const apiGateway = new RestApi(scope, 'RestApi', {
			restApiName: `${scope.stack}-${scope.stage}-${scope.app}`,
		});

		if (scope.stage === 'PROD' && !props.monitoring.noMonitoring) {
			new SrApiGateway5xxAlarm(scope, {
				errorImpact: props.monitoring.errorImpact,
				restApi: apiGateway,
				lambdaFunctionNames: [],
			});
		}

		apiGateway.root
			.resourceForPath('/')
			.addMethod('POST', sendMessageIntegration, {
				methodResponses: [
					{
						statusCode: '200',
					},
				],
			});
	}
}
