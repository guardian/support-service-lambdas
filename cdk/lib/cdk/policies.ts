import type { GuStack } from '@guardian/cdk/lib/constructs/core';
import {
	GuAllowPolicy,
	GuGetS3ObjectsPolicy,
} from '@guardian/cdk/lib/constructs/iam';
import { Fn } from 'aws-cdk-lib';

class AllowGetSecretValuePolicy extends GuAllowPolicy {
	constructor(scope: GuStack, id: string, key: string) {
		super(scope, id, {
			actions: ['secretsmanager:GetSecretValue'],
			resources: [
				`arn:aws:secretsmanager:${scope.region}:${scope.account}:secret:${scope.stage}/${key}`,
			],
		});
	}
}

export type SrQueueName =
	| `braze-emails`
	| 'supporter-product-data'
	| 'product-switch-salesforce-tracking';

export class AllowSqsSendPolicy extends GuAllowPolicy {
	constructor(
		scope: GuStack,
		queuePrefix: SrQueueName,
		...additionalQueuePrefixes: readonly SrQueueName[]
	) {
		const queuePrefixes = [queuePrefix, ...additionalQueuePrefixes];
		const resources = queuePrefixes.map(
			(queuePrefix) =>
				`arn:aws:sqs:${scope.region}:${scope.account}:${queuePrefix}-${scope.stage}`,
		);
		super(scope, 'SQS send policy', {
			actions: ['sqs:GetQueueUrl', 'sqs:SendMessage'],
			resources,
		});
	}
}

export class AllowS3CatalogReadPolicy extends GuGetS3ObjectsPolicy {
	constructor(scope: GuStack) {
		super(scope, 'S3 catalog read policy', {
			bucketName: 'gu-zuora-catalog',
			paths: [`PROD/Zuora-${scope.stage}/*`],
		});
	}
}

export class AllowZuoraOAuthSecretsPolicy extends AllowGetSecretValuePolicy {
	constructor(scope: GuStack) {
		super(
			scope,
			'Zuora OAuth Secrets Manager policy',
			'Zuora-OAuth/SupportServiceLambdas-*',
		);
	}
}

export class AllowSupporterProductDataQueryPolicy extends GuAllowPolicy {
	constructor(scope: GuStack) {
		super(scope, 'SupporterProductDataTable query access', {
			actions: ['dynamodb:Query'],
			resources: [
				Fn.importValue(
					`supporter-product-data-tables-${scope.stage}-SupporterProductDataTable`,
				),
			],
		});
	}
}
