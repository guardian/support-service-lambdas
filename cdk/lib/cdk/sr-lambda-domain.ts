import type { GuStack } from '@guardian/cdk/lib/constructs/core';
import type { IRestApi } from 'aws-cdk-lib/aws-apigateway';
import { CfnBasePathMapping, CfnDomainName } from 'aws-cdk-lib/aws-apigateway';
import { CfnRecordSet } from 'aws-cdk-lib/aws-route53';
import { Construct } from 'constructs';

interface GuDomainProps {
    subdomain: string; // e.g. "metric-push-api", "mparticle-api"
    stage: string;
    restApi: IRestApi;
    apiDomain: 'support' | 'membership';
}

function generateSafeId(input: string): string {
    return input
        .split('-')
        .map(
            (word) =>
                word.charAt(0).toUpperCase() + word.slice(1).toLocaleLowerCase(),
        )
        .join('');
}

export class SrLambdaDomain extends Construct {
    constructor(scope: GuStack, props: GuDomainProps) {
        const safeId = generateSafeId(props.subdomain);
        super(scope, `${safeId}Domain`);

        let apiDomain, certificateId;
        switch (props.apiDomain) {
            case "support":
                apiDomain = "support.guardianapis.com";
                certificateId = "b384a6a0-2f54-4874-b99b-96eeff96c009";
                break;
            case "membership":
                apiDomain = "membership.guardianapis.com";
                certificateId = "c1efc564-9ff8-4a03-be48-d1990a3d79d2";
                break;
        }

        const domainNameString = `${props.subdomain}-${props.stage.toLowerCase()}.${apiDomain}`;

        const domainName = new CfnDomainName(this, `${safeId}DomainName`, {
            regionalCertificateArn: `arn:aws:acm:${scope.region}:${scope.account}:certificate/${certificateId}`,
            domainName: domainNameString,
            endpointConfiguration: { types: ['REGIONAL'] },
        });

        new CfnBasePathMapping(this, `${safeId}BasePathMapping`, {
            restApiId: props.restApi.restApiId,
            domainName: domainName.ref,
            stage: props.restApi.deploymentStage.stageName,
        });

        const dnsRecord = new CfnRecordSet(this, `${safeId}DnsRecord`, {
            name: domainNameString,
            type: 'CNAME',
            comment: `CNAME for ${props.subdomain} API ${props.stage}`,
            hostedZoneName: `${apiDomain}.`,
            ttl: '120',
            resourceRecords: [domainName.attrRegionalDomainName],
        });
        dnsRecord.overrideLogicalId(`${safeId}DnsRecord`);
    }
}
