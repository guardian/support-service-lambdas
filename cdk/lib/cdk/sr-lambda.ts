import type { Identity } from '@guardian/cdk/lib/constructs/core';
import type { GuFunctionProps } from '@guardian/cdk/lib/constructs/lambda';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import { Duration } from 'aws-cdk-lib';
import { LoggingFormat } from 'aws-cdk-lib/aws-lambda';
import { nodeVersion } from '../node-version';
import type { SrStack } from './sr-stack';

export type SrLambdaProps = {
	nameSuffix?: string; // when you have multiple lambdas in an app
};

type DefaultProps = ReturnType<typeof getLambdaDefaultProps>;
type GuLambdaOverrides = Omit<GuFunctionProps, keyof DefaultProps> &
	Partial<DefaultProps>;

export function getNameWithStage(
	identity: Identity,
	nameSuffix: string | undefined,
) {
	return `${identity.app}${nameSuffix ? '-' + nameSuffix : ''}-${identity.stage}`;
}

export function getLambdaDefaultProps(
	scope: Identity,
	nameSuffix: string | undefined,
) {
	return {
		app: scope.app,
		functionName: getNameWithStage(scope, nameSuffix),
		fileName: `${scope.app}.zip`,
		handler: 'index.handler',
		runtime: nodeVersion,
		loggingFormat: LoggingFormat.TEXT,
		memorySize: 1024,
		timeout: Duration.seconds(15),
		environment: {
			NODE_OPTIONS: '--enable-source-maps',
		} as Record<string, string>,
	};
}

export class SrLambda extends GuLambdaFunction {
	constructor(
		scope: SrStack,
		id: string,
		lambdaOverrides: GuLambdaOverrides,
		props: SrLambdaProps,
	) {
		const defaultProps = getLambdaDefaultProps(scope, props.nameSuffix);
		const finalProps = {
			...defaultProps,
			...lambdaOverrides,
			environment: {
				...defaultProps.environment,
				...lambdaOverrides.environment,
			},
		};

		super(scope, id, finalProps);
	}
}
