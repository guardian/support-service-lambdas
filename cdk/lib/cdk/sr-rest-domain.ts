import { GuCname } from '@guardian/cdk/lib/constructs/dns';
import { Duration } from 'aws-cdk-lib';
import type { LambdaRestApi } from 'aws-cdk-lib/aws-apigateway';
import { CfnBasePathMapping, CfnDomainName } from 'aws-cdk-lib/aws-apigateway';
import { CfnRecordSet } from 'aws-cdk-lib/aws-route53';
import { certForStack } from '../constants';
import type { SrStack } from './sr-stack';

export class SrRestDomain {
	readonly dnsRecord: CfnRecordSet;
	readonly cfnDomainName: CfnDomainName;
	readonly basePathMapping: CfnBasePathMapping;
	constructor(
		scope: SrStack,
		api: LambdaRestApi,
		suffixProdDomain: boolean = false,
		publicDomain: boolean = false, // if setting it to true, you have to add a fastly configuration for it
	) {
		const app = scope.app;

		const isProd = scope.stage === 'PROD';
		const cert = certForStack[scope.stack];
		// ---- DNS ---- //
		const certificateArn = `arn:aws:acm:eu-west-1:${scope.account}:certificate/${cert.certificateId}`;
		const domainName = `${app}${isProd && !suffixProdDomain ? '' : '-' + scope.stage.toLowerCase()}.${cert.domainName}`;

		this.cfnDomainName = new CfnDomainName(scope, 'DomainName', {
			domainName,
			regionalCertificateArn: certificateArn,
			endpointConfiguration: {
				types: ['REGIONAL'],
			},
		});

		this.basePathMapping = new CfnBasePathMapping(scope, 'BasePathMapping', {
			domainName: this.cfnDomainName.ref,
			restApiId: api.restApiId,
			stage: api.deploymentStage.stageName,
		});

		this.dnsRecord = new CfnRecordSet(scope, 'DNSRecord', {
			name: domainName,
			type: 'CNAME',
			hostedZoneId: cert.hostedZoneId,
			ttl: '120',
			resourceRecords: [this.cfnDomainName.attrRegionalDomainName],
		});

		if (publicDomain) {
			const domainName =
				app +
				(scope.stage === 'PROD'
					? `.guardianapis.com`
					: `.code.dev-guardianapis.com`);
			new GuCname(scope, `NS1 DNS entry for ${domainName}`, {
				app,
				domainName,
				ttl: Duration.hours(1),
				resourceRecord: 'dualstack.guardian.map.fastly.net',
			});
		}
	}
}
