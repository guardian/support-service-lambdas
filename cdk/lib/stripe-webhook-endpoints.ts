import path from 'path';
import { GuApiGatewayWithLambdaByPath } from '@guardian/cdk';
import { GuAlarm } from '@guardian/cdk/lib/constructs/cloudwatch';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import type { App } from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { CfnBasePathMapping, CfnDomainName } from 'aws-cdk-lib/aws-apigateway';
import { ComparisonOperator, Metric } from 'aws-cdk-lib/aws-cloudwatch';
import { Effect, Policy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { LoggingFormat, Runtime } from 'aws-cdk-lib/aws-lambda';
import { CfnRecordSet } from 'aws-cdk-lib/aws-route53';
import { CfnInclude } from 'aws-cdk-lib/cloudformation-include';

export interface StripeWebhookEndpointsProps extends GuStackProps {
	stack: string;
	stage: string;
	certificateId: string;
	domainName: string;
	hostedZoneId: string;
}

export class StripeWebhookEndpoints extends GuStack {
	constructor(scope: App, id: string, props: StripeWebhookEndpointsProps) {
		super(scope, id, props);

		const app = 'stripe-webhook-endpoints';

		// ---- Existing CFN template ---- //
		const yamlTemplateFilePath = path.join(
			__dirname,
			'../..',
			'handlers/stripe-webhook-endpoints/cfn.yaml',
		);
		new CfnInclude(this, 'YamlTemplate', {
			templateFile: yamlTemplateFilePath,
		});

		// ---- API-triggered lambda functions ---- //

		const paymentIntentIssuesLambda = new GuLambdaFunction(
			this,
			'payment-intent-issues-cdk-lambda',
			{
				app: app,
				description:
					'A lambda for handling payment intent issues (cancellation, failure, action required)',
				functionName: `stripe-payment-intent-issues-cdk-${this.stage}`,
				loggingFormat: LoggingFormat.TEXT,
				fileName: `${app}.jar`,
				handler: 'com.gu.paymentIntentIssues.Lambda::handler',
				runtime: Runtime.JAVA_21,
				memorySize: 1536,
				timeout: Duration.seconds(300),
				environment: {
					App: app,
					Stack: this.stack,
					Stage: this.stage,
				},
			},
		);

		const customerUpdatedLambda = new GuLambdaFunction(
			this,
			'customer-updated-cdk-lambda',
			{
				app: app,
				description: 'A lambda for handling customer updates',
				functionName: `stripe-customer-updated-cdk-${this.stage}`,
				loggingFormat: LoggingFormat.TEXT,
				fileName: `${app}.jar`,
				handler: 'com.gu.stripeCardUpdated.Lambda::apply',
				runtime: Runtime.JAVA_21,
				memorySize: 1536,
				timeout: Duration.seconds(900),
				environment: {
					App: app,
					Stack: this.stack,
					Stage: this.stage,
				},
			},
		);

		// Wire up the API
		// ---- API gateway ---- //
		const stripeWebhookEndpointsApi = new GuApiGatewayWithLambdaByPath(this, {
			app: 'stripe-webhook-endpoints',
			targets: [
				{
					path: '/payment-intent-issue',
					httpMethod: 'POST',
					lambda: paymentIntentIssuesLambda,
				},
				{
					path: '/customer-updated',
					httpMethod: 'POST',
					lambda: customerUpdatedLambda,
				},
			],
			// Create an alarm
			monitoringConfiguration: {
				snsTopicName: 'conversion-dev',
				http5xxAlarm: {
					tolerated5xxPercentage: 1,
				},
			},
		});

		// ---- Alarms ---- //
		const alarmName = (shortDescription: string) =>
			`Stripe Webhook Endpoints CDK ${this.stage} ${shortDescription}`;

		const alarmDescription = (description: string) =>
			`Impact - ${description}. Follow the process in https://docs.google.com/document/d/1_3El3cly9d7u_jPgTcRjLxmdG2e919zCLvmcFCLOYAk/edit`;

		new GuAlarm(this, 'ApiGateway4XXAlarmCDK', {
			app,
			alarmName: alarmName('API gateway 4XX response'),
			alarmDescription: alarmDescription(
				'Stripe Webhook Endpoints received an invalid request',
			),
			evaluationPeriods: 1,
			threshold: 1,
			snsTopicName: 'conversion-dev',
			actionsEnabled: this.stage === 'PROD',
			comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
			metric: new Metric({
				metricName: '4XXError',
				namespace: 'AWS/ApiGateway',
				statistic: 'Sum',
				period: Duration.seconds(300),
				dimensionsMap: {
					ApiName: `${app}-${this.stage}`,
				},
			}),
		});

		new GuAlarm(this, 'ApiGateway5XXAlarmCDK', {
			app,
			alarmName: alarmName('API gateway 5XX error'),
			alarmDescription: `stripe-webhook-endpoints-${this.stage} exceeded 1% 5XX error rate`,
			evaluationPeriods: 1,
			threshold: 1,
			actionsEnabled: this.stage === 'PROD',
			snsTopicName: 'conversion-dev',
			comparisonOperator: ComparisonOperator.GREATER_THAN_THRESHOLD,
			metric: new Metric({
				metricName: '5XXError',
				namespace: 'AWS/ApiGateway',
				statistic: 'Sum',
				period: Duration.seconds(60),
				dimensionsMap: {
					ApiName: `${app}-${this.stage}`,
				},
			}),
		});

		// ---- DNS ---- //
		const certificateArn = `arn:aws:acm:${this.region}:${this.account}:certificate/${props.certificateId}`;

		const cfnDomainName = new CfnDomainName(
			this,
			'StripeWebhookEndpointsDomainName',
			{
				domainName: props.domainName,
				regionalCertificateArn: certificateArn,
				endpointConfiguration: {
					types: ['REGIONAL'],
				},
			},
		);

		new CfnBasePathMapping(this, 'BasePathMapping', {
			domainName: cfnDomainName.ref,
			// Uncomment the lines below to reroute traffic to the new API Gateway instance
			restApiId: stripeWebhookEndpointsApi.api.restApiId,
			stage: stripeWebhookEndpointsApi.api.deploymentStage.stageName,
			// Uncomment the lines below to reroute traffic to the old (existing) API Gateway instance
			// restApiId: yamlDefinedResources.getResource("ServerlessRestApi").ref,
			// stage: props.stage,
		});

		new CfnRecordSet(this, 'DNSRecord', {
			name: props.domainName,
			type: 'CNAME',
			hostedZoneId: props.hostedZoneId,
			ttl: '120',
			resourceRecords: [cfnDomainName.attrRegionalDomainName],
		});

		// ---- Apply policies ---- //
		const ssmInlinePolicy: Policy = new Policy(this, 'SSM inline policy', {
			statements: [
				new PolicyStatement({
					effect: Effect.ALLOW,
					actions: ['ssm:GetParametersByPath'],
					resources: [
						`arn:aws:ssm:${this.region}:${this.account}:parameter/${props.stage}/membership/payment-intent-issues/*`,
					],
				}),
			],
		});

		const s3InlinePolicyForPaymentIntentIssues: Policy = new Policy(
			this,
			'S3 inline policy',
			{
				statements: [
					new PolicyStatement({
						effect: Effect.ALLOW,
						actions: ['s3:GetObject'],
						resources: ['arn:aws:s3::*:membership-dist/*'],
					}),
				],
			},
		);

		const s3InlinePolicyForCustomerUpdated: Policy = new Policy(
			this,
			'S3 inline policy For Customer Updated lambda',
			{
				statements: [
					new PolicyStatement({
						effect: Effect.ALLOW,
						actions: ['s3:GetObject'],
						resources: [
							`arn:aws:s3::*:membership-dist/*`,
							`arn:aws:s3:::gu-reader-revenue-private/membership/support-service-lambdas/${props.stage}/zuoraRest-${props.stage}.*.json`,
							`arn:aws:s3:::gu-reader-revenue-private/membership/support-service-lambdas/${props.stage}/trustedApi-${props.stage}.*.json`,
							`arn:aws:s3:::gu-reader-revenue-private/membership/support-service-lambdas/${props.stage}/stripe-${props.stage}.*.json`,
						],
					}),
				],
			},
		);

		paymentIntentIssuesLambda.role?.attachInlinePolicy(ssmInlinePolicy);
		paymentIntentIssuesLambda.role?.attachInlinePolicy(
			s3InlinePolicyForPaymentIntentIssues,
		);
		customerUpdatedLambda.role?.attachInlinePolicy(ssmInlinePolicy);
		customerUpdatedLambda.role?.attachInlinePolicy(
			s3InlinePolicyForCustomerUpdated,
		);
	}
}
