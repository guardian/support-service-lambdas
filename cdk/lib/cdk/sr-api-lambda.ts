import { GuApiLambda, type GuApiLambdaProps } from '@guardian/cdk';
import type { Identity } from '@guardian/cdk/lib/constructs/core';
import type { GuPolicy } from '@guardian/cdk/lib/constructs/iam/policies/base-policy';
import { Duration } from 'aws-cdk-lib';
import { ApiKeySourceType } from 'aws-cdk-lib/aws-apigateway';
import { ComparisonOperator, Metric } from 'aws-cdk-lib/aws-cloudwatch';
import type { SrLambdaProps } from './sr-lambda';
import { getLambdaDefaultProps, getNameWithStage } from './sr-lambda';
import type { SrLambdaAlarmProps } from './sr-lambda-alarm';
import { SrLambdaAlarm } from './sr-lambda-alarm';
import type { SrRestDomainProps } from './sr-rest-domain';
import { SrRestDomain } from './sr-rest-domain';
import type { SrStack } from './sr-stack';

function getApiLambdaDefaultProps(scope: Identity, props: SrApiLambdaProps) {
	return {
		...getLambdaDefaultProps(scope, props.nameSuffix),
		timeout: Duration.seconds(300),
		monitoringConfiguration: {
			noMonitoring: true, // we don't use the standard GuCDK alarms due to low traffic
		} as GuApiLambdaProps['monitoringConfiguration'],
		api: {
			id: getNameWithStage(scope, props.nameSuffix),
			restApiName: getNameWithStage(scope, props.nameSuffix),
			description: props.apiDescriptionOverride ?? 'API Gateway created by CDK',
			proxy: false, // add proxy method later, to allow public paths
			deployOptions: {
				stageName: scope.stage,
			},
			...(props.isPublic
				? {}
				: {
						apiKeySourceType: ApiKeySourceType.HEADER,
						defaultMethodOptions: {
							apiKeyRequired: true,
						},
					}),
		},
	};
}

type ApiDefaultProps = ReturnType<typeof getApiLambdaDefaultProps>;
type SrApiLambdaOverrides = Omit<GuApiLambdaProps, keyof ApiDefaultProps> &
	Partial<ApiDefaultProps>;
type SrApiLambdaProps = SrLambdaProps & {
	lambdaOverrides: SrApiLambdaOverrides;
	apiDescriptionOverride?: string;
	isPublic?: boolean;
	errorImpact: string;
	alarmOverrides?: Partial<SrLambdaAlarmProps>;
	srRestDomainOverrides?: SrRestDomainProps;
};

export class SrApiLambda extends GuApiLambda {
	readonly domain: SrRestDomain;
	constructor(scope: SrStack, props: SrApiLambdaProps) {
		const defaultProps = getApiLambdaDefaultProps(scope, props);
		const deprecatedVars = {
			// for some reason we often use these instead of the upper case STACK/STAGE/APP added by GuCDK
			// https://github.com/guardian/cdk/blob/5569c749211b518001666cffb558fe403ff0539c/src/constructs/lambda/lambda.ts#L134-L138
			App: scope.app,
			Stack: scope.stack,
			Stage: scope.stage,
		};
		const finalProps = {
			...defaultProps,
			...props.lambdaOverrides,
			environment: {
				...defaultProps.environment,
				...deprecatedVars,
				...props.lambdaOverrides.environment,
			},
		};

		super(scope, `${scope.app}-lambda`, finalProps);

		// by doing these explicitly rather than using proxy:true, we can later add specific public resources
		this.api.root.addMethod('ANY');
		this.api.root.addResource('{proxy+}').addMethod('ANY');

		if (!props.isPublic) {
			const usagePlan = this.api.addUsagePlan('UsagePlan', {
				name: getNameWithStage(scope, props.nameSuffix),
				description: 'REST endpoints for ' + scope.app,
				apiStages: [
					{
						stage: this.api.deploymentStage,
						api: this.api,
					},
				],
			});

			// create api key
			const apiKey = this.api.addApiKey(`${scope.app}-key-${scope.stage}`, {
				apiKeyName: `${scope.app}-key-${scope.stage}`,
			});

			// associate api key to plan
			usagePlan.addApiKey(apiKey);
		}

		this.domain = new SrRestDomain(
			scope,
			this.api,
			props.srRestDomainOverrides,
		);

		if (
			scope.stage === 'PROD' &&
			finalProps.monitoringConfiguration.noMonitoring
		) {
			new SrLambdaAlarm(scope, 'ApiGateway5XXAlarmCDK', {
				app: scope.app,
				alarmName: getNameWithStage(scope, props.nameSuffix) + ' 5XX errors',
				alarmDescription:
					scope.app +
					' returned a 5XX response search the logs below for "error" for more information. Impact: ' +
					props.errorImpact,
				evaluationPeriods: 1,
				threshold: 1,
				lambdaFunctionNames: this.functionName,
				comparisonOperator:
					ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
				metric: new Metric({
					metricName: '5XXError',
					namespace: 'AWS/ApiGateway',
					statistic: 'Sum',
					period: Duration.seconds(60),
					dimensionsMap: {
						ApiName: getNameWithStage(scope, props.nameSuffix),
					},
				}),
				...props.alarmOverrides,
			});
		}
	}

	addPolicies(...policies: GuPolicy[]) {
		policies.forEach((p) => this.role!.attachInlinePolicy(p));
	}

	addPublicPath(path: string) {
		const publicResource = this.api.root.addResource(path);
		publicResource.addMethod('GET', undefined, {
			apiKeyRequired: false,
		});
		publicResource.addResource('{proxy+}').addMethod('GET', undefined, {
			apiKeyRequired: false,
		});
	}
}
