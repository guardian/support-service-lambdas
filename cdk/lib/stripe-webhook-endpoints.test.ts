import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { stripeWebhookEndpointsCodeProps, stripeWebhookEndpointsProdProps } from "../bin/cdk";
import { StripeWebhookEndpoints } from "./stripe-webhook-endpoints";

describe("The StripeWebhookEndpoints stack", () => {
    it("matches the snapshot", () => {
        const app = new App();
        // const stack = new StripeWebhookEndpoints(app, "StripeWebhookEndpoints", { stack: "membership", stage: "TEST" });

        const codeStack = new StripeWebhookEndpoints(app, "Stripe-Webhook-Endpoints-CODE",stripeWebhookEndpointsCodeProps);
        const prodStack = new StripeWebhookEndpoints(app, "Stripe-Webhook-Endpoints-PROD", stripeWebhookEndpointsProdProps);

        // const template = Template.fromStack(stack);
        // expect(template.toJSON()).toMatchSnapshot();

        expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
        expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();

    });
});