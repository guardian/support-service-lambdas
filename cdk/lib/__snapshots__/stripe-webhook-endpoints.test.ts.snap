// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The StripeWebhookEndpoints stack matches the snapshot 1`] = `
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Conditions": {
    "IsProd": {
      "Fn::Equals": [
        {
          "Ref": "Stage",
        },
        "PROD",
      ],
    },
  },
  "Description": "Endpoints to handle stripe webhooks",
  "Metadata": {
    "gu:cdk:constructs": [],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "App": {
      "Default": "stripe-webhook-endpoints",
      "Description": "Endpoints to handle stripe webhooks",
      "Type": "String",
    },
    "DeployBucket": {
      "Default": "membership-dist",
      "Description": "Bucket to copy files to",
      "Type": "String",
    },
    "Stack": {
      "Default": "membership",
      "Description": "Stack name",
      "Type": "String",
    },
    "Stage": {
      "AllowedValues": [
        "CODE",
        "PROD",
      ],
      "Description": "Set by RiffRaff on each deploy",
      "Type": "String",
    },
  },
  "Resources": {
    "CustomerUpdatedLambda": {
      "Properties": {
        "CodeUri": {
          "Bucket": {
            "Ref": "DeployBucket",
          },
          "Key": {
            "Fn::Sub": "\${Stack}/\${Stage}/\${App}/\${App}.jar",
          },
        },
        "Description": "A lambda for handling customer updates",
        "Environment": {
          "Variables": {
            "App": "stripe-customer-updated",
            "Stack": {
              "Ref": "Stack",
            },
            "Stage": {
              "Ref": "Stage",
            },
          },
        },
        "Events": {
          "AcquisitionEvent": {
            "Properties": {
              "Method": "post",
              "Path": "/customer-updated",
              "RequestParameters": [
                "method.request.querystring.apiToken",
              ],
            },
            "Type": "Api",
          },
        },
        "FunctionName": {
          "Fn::Sub": "stripe-customer-updated-\${Stage}",
        },
        "Handler": "com.gu.stripeCardUpdated.Lambda::apply",
        "MemorySize": 1536,
        "Policies": [
          "AWSLambdaBasicExecutionRole",
          {
            "Statement": {
              "Action": [
                "ssm:GetParametersByPath",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Sub": "arn:aws:ssm:\${AWS::Region}:\${AWS::AccountId}:parameter/\${Stage}/membership/payment-intent-issues",
                },
              ],
            },
          },
          {
            "Statement": [
              {
                "Action": "s3:GetObject",
                "Effect": "Allow",
                "Resource": [
                  "arn:aws:s3::*:membership-dist/*",
                  {
                    "Fn::Sub": "arn:aws:s3:::gu-reader-revenue-private/membership/support-service-lambdas/\${Stage}/zuoraRest-\${Stage}.*.json",
                  },
                  {
                    "Fn::Sub": "arn:aws:s3:::gu-reader-revenue-private/membership/support-service-lambdas/\${Stage}/trustedApi-\${Stage}.*.json",
                  },
                  {
                    "Fn::Sub": "arn:aws:s3:::gu-reader-revenue-private/membership/support-service-lambdas/\${Stage}/stripe-\${Stage}.*.json",
                  },
                ],
              },
            ],
          },
        ],
        "Runtime": "java21",
        "Tags": {
          "Stack": "membership",
          "Stage": "TEST",
          "gu:cdk:version": "TEST",
          "gu:repo": "guardian/support-service-lambdas",
        },
        "Timeout": 900,
      },
      "Type": "AWS::Serverless::Function",
    },
    "PaymentIntentIssuesLambda": {
      "Properties": {
        "CodeUri": {
          "Bucket": {
            "Ref": "DeployBucket",
          },
          "Key": {
            "Fn::Sub": "\${Stack}/\${Stage}/\${App}/\${App}.jar",
          },
        },
        "Description": "A lambda for handling payment intent issues (cancellation, failure, action required)",
        "Environment": {
          "Variables": {
            "App": "payment-intent-issues",
            "Stack": {
              "Ref": "Stack",
            },
            "Stage": {
              "Ref": "Stage",
            },
          },
        },
        "Events": {
          "AcquisitionEvent": {
            "Properties": {
              "Method": "post",
              "Path": "/payment-intent-issue",
            },
            "Type": "Api",
          },
        },
        "FunctionName": {
          "Fn::Sub": "stripe-payment-intent-issues-\${Stage}",
        },
        "Handler": "com.gu.paymentIntentIssues.Lambda::handler",
        "MemorySize": 512,
        "Policies": [
          "AWSLambdaBasicExecutionRole",
          {
            "Statement": {
              "Action": [
                "ssm:GetParametersByPath",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Sub": "arn:aws:ssm:\${AWS::Region}:\${AWS::AccountId}:parameter/\${Stage}/membership/payment-intent-issues",
                },
              ],
            },
          },
          {
            "Statement": [
              {
                "Action": "s3:GetObject",
                "Effect": "Allow",
                "Resource": [
                  "arn:aws:s3::*:membership-dist/*",
                ],
              },
            ],
          },
        ],
        "Runtime": "java21",
        "Tags": {
          "Stack": "membership",
          "Stage": "TEST",
          "gu:cdk:version": "TEST",
          "gu:repo": "guardian/support-service-lambdas",
        },
        "Timeout": 300,
      },
      "Type": "AWS::Serverless::Function",
    },
  },
  "Transform": "AWS::Serverless-2016-10-31",
}
`;
